components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadStorage.kt

```diff
@@ -0,0 +1,89 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.downloads
+
+import android.content.Context
+import androidx.annotation.VisibleForTesting
+import androidx.paging.DataSource
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.map
+import mozilla.components.browser.state.state.content.DownloadState
+import mozilla.components.feature.downloads.db.DownloadsDatabase
+import mozilla.components.feature.downloads.db.toDownloadEntity
+
+/**
+ * A storage implementation for organizing download.
+ */
+class DownloadStorage(context: Context) {
+
+    @VisibleForTesting
+    internal var database: Lazy<DownloadsDatabase> = lazy { DownloadsDatabase.get(context) }
+
+    private val downloadDao by lazy { database.value.downloadDao() }
+
+    /**
+     * Adds a new [download].
+     */
+    suspend fun add(download: DownloadState) {
+        downloadDao.insert(download.toDownloadEntity())
+    }
+
+    /**
+     * Returns a [Flow] list of all the [DownloadState] instances.
+     */
+    fun getDownloads(): Flow<List<DownloadState>> {
+        return downloadDao.getDownloads().map { list ->
+            list.map { entity -> entity.toDownloadState() }
+        }
+    }
+
+    /**
+     * Returns all saved [DownloadState] instances as a [DataSource.Factory].
+     */
+    fun getDownloadsPaged(): DataSource.Factory<Int, DownloadState> = downloadDao
+        .getDownloadsPaged()
+        .map { entity ->
+            entity.toDownloadState()
+        }
+
+    /**
+     * Removes the given [download].
+     */
+    suspend fun remove(download: DownloadState) {
+        downloadDao.delete(download.toDownloadEntity())
+    }
+
+    /**
+     * Update the given [download].
+     */
+    suspend fun update(download: DownloadState) {
+        downloadDao.update(download.toDownloadEntity())
+    }
+
+    /**
+     * Removes all the downloads.
+     */
+    suspend fun removeAllDownloads() {
+        downloadDao.deleteAllDownloads()
+    }
+
+    companion object {
+        /**
+         * Takes two [DownloadState] objects and the determine if they are the same, be aware this
+         * only takes into considerations fields that are being stored,
+         * not all the field on [DownloadState] are stored.
+         */
+        fun areTheSame(first: DownloadState, second: DownloadState): Boolean {
+            return first.id == second.id &&
```

Comment:
Now that we have a unique ID, comparing the ID should be good enough here. This would hide a bug if we had two downloads with the same ID but different file name.

Reply:
We are using this function to validate if something has changed, between updates. We need to verify for each field that the storage cares about, because we don't overwhelm the storage, if there is not new info, maybe `DownloadStorage` is not a good place. What do you think?


```kotlin
            is DownloadAction.UpdateDownloadAction -> {
                val updated = action.download
                store.state.downloads[updated.id]?.let { old ->
                    // To not overwhelm the storage, we only send updates that are relevant,
                    // we only care about properties, that we are stored on the storage.
                    if (!DownloadStorage.areTheSame(old, updated)) {
                        scope.launch {
                            downloadStorage.update(action.download)
                        }
                        logger.debug("Updated download ${action.download.fileName} on the storage")
                    }
                }
            }
```

Reply:
OK, see comment below. I think we can just choose a clearer name and keep everything as-is otherwise.

Reply:
Any suggestions for the new name?

Reply:
What about `DownloadStorage.equals(a, b)` ?